### PR TITLE
Add `torch.logcumsumexp`

### LINKED ISSUE #26411: LogCumsumExp
## 🚀 Feature
Add numerically stable cumulative logsumexp function. Also we have associated PR on `cummax` that is needed for numerically stable implementation (https://github.com/pytorch/pytorch/issues/20240).
## Motivation
This is useful when computing sum of probabilities and have different applications.

## Pitch
Torch has `cumsum` and `cumprod` so I suggest `logcumsumexp` to be added.

## Alternatives
Current workaround for me and for people that need it can be written as follows (it is quite fast, only overhead for converting between numpy/torch):
```
import numpy as np
import torch

def cummax(x, dim):
    x_np = x.detach().cpu().numpy()
    ret = np.maximum.accumulate(x_np, axis=dim)
    return torch.from_numpy(ret).to(x)


def logcumsumexp(x, dim=-1):
    if (dim != -1) or (dim != x.ndimension() - 1):
        x = x.transpose(dim, -1)

    init_size = x.size()
    last_dim_size = init_size[-1]
    x_resized = x.contiguous().view(-1, last_dim_size)
    d1, d2 = x_resized.size()
    x_cummax = cummax(x_resized, -1).view(d1, d2, 1)
    x_expand = x_resized.unsqueeze(1).expand(d1, d2, last_dim_size)
    mask = torch.tril(torch.ones(last_dim_size, last_dim_size)).unsqueeze(0)
    ret = torch.log(torch.sum(torch.exp(x_expand - x_cummax) * mask, dim=-1)) + x_cummax.view(d1, d2)
    ret = ret.view(*init_size)

    if (dim != -1) or (dim != x.ndimension() - 1):
        ret = ret.transpose(-1, dim)

    return ret
```

UPDATE: code above is not always numerically stable. Still most numerically stable, but slow way to compute `logcumsumexp` is use for-loop:

```
def logcumsumexp(x, dim):
    # slow implementation, but ok for now
    if (dim != -1) or (dim != x.ndimension() - 1):
        x = x.transpose(dim, -1)

    out = []
    for i in range(1, x.size(-1) + 1):
        out.append(torch.logsumexp(x[..., :i], dim=-1, keepdim=True))
    out = torch.cat(out, dim=-1)

    if (dim != -1) or (dim != x.ndimension() - 1):
        out = out.transpose(-1, dim)
    return out
```

### PR DESCRIPTION
**Pull request by kshitij12345**
Creating new PR as I am unable to push to @pandeykartikey 's branch as I don't have the permissions.

Closes #26411 

Based on #32876 Thanks @pandeykartikey for starting this out.

Have addressed the comments.

@anjali411 @agadetsky @albanD

### CODE REVIEW

**Comment by tridao** [comment]
I don't think the current implementation is numerically stable. The original issue #26411 suggests using cummax to reduce numerical error, but that has quadratic complexity (see discussion in #32876).
I still think it's better to implement it the way Tensorflow does it: as a prefix scan of the binary operation log_add_exp.

**Comment by kshitij12345** [comment]
@anjali411 @albanD Could you please review the PR. Also I don't have the rights to re-run the failed pipeline (whose failure is unrelated to the PR).

@tridao I get the point . But I am not sure that I am quite familiar with the codebase. I saw the reference for `cumsum` https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
and https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathScan.cu ( CUDA Implementation which I don't think should be used as reference). 
Would be great if I can get some pointers.

**Comment by tridao** [comment]
Agreed, those seem to be where cumsum is implemented.
I imagine one just needs to replace the addition (x + y) of cumsum with the operation log_add_exp(x, y)(implemented in a stable way with max and min, e.g. https://www.tensorflow.org/api_docs/python/tf/math/cumulative_logsumexp). 

Re: CUDA implementation in THC: maybe cumsum will eventually be ported from THC to Aten? I also found cummax implementation in Aten (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/ReduceOpsKernel.cu). One would replace max(x, y) with log_add_exp. The cummax implementation also needs the indices but I don't think logcumsumexp will need those.

**Comment by kshitij12345** [comment]
@tridao Thanks for the reference and help. 

Have added the kernels based on `log_add_exp` operator. 

Please review @tridao @anjali411 @albanD

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (tridao) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:122]: Minor point: maybe you should use the log1p function (log (1 + x)) instead of log.

File: aten/src/ATen/native/cuda/ReduceOpsKernel.cu
Comment (tridao) [inline aten/src/ATen/native/cuda/ReduceOpsKernel.cu:656]: Minor point: maybe you should use the log1p function (log (1 + x)) instead of log.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp]: Makes sense. Thanks.

Resolved.

File: aten/src/ATen/native/cuda/ReduceOpsKernel.cu
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/ReduceOpsKernel.cu]: Makes sense. Thanks.

Resolved.

**Comment by kshitij12345** [comment]
@anjali411 @albanD Would be great if you can review the PR.

**Comment by kshitij12345** [comment]
@anjali411 @albanD Gentle ping for reminder :)

**Comment by agadetsky** [comment]
@anjali411 @albanD Gentle ping

File: aten/src/ATen/native/ReduceOps.cpp
Comment (anjali411) [inline aten/src/ATen/native/ReduceOps.cpp:192]: `return _logcumsumexp_out_cpu(result, self, dim);`

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (anjali411) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:112]: can we also add this for complex dtypes?
`AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES`

Comment (anjali411) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:112]: ahh I guess we can't extend this to complex if we are using `max` and `min` ops in the implementation

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (anjali411) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:127]: why should the init value be -inf?

**Review by anjali411 — APPROVED** [review]
LGTM overall :) let's rebase to resolve the merge conflicts

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:127]: For `cumsum` part, we need initial value of accumulator to be `0`.

`exp(-inf)` is `0`. So, that is used as the initial value.

File: aten/src/ATen/native/ReduceOps.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/ReduceOps.cpp:192]: Noted.

**Comment by anjali411** [comment]
@kshitij12345 can you also benchmark cpu and cuda performance?

**Comment by kshitij12345** [comment]
Benchmark with same build settings on same system.
gcc : version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
CUDA : 10.1
GPU : 1050ti

## CUDA Timings:

```python
import timeit

for n, t in [(50, 20000),
             (100, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.logcumsumexp(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.logcumsumexp(a, 0);torch.logcumsumexp(a, 1); torch.cuda.synchronize()',
                              setup=f'import torch; a=torch.randn(({n}, {n}), dtype={dtype}, device="cuda")',
                              number=t))
```

```
torch.logcumsumexp(a) a.numel() == 50 for 20000 times torch.half
1.085265507994336
torch.logcumsumexp(a) a.numel() == 50 for 20000 times torch.float
0.8086711470023147
torch.logcumsumexp(a) a.numel() == 50 for 20000 times torch.double
3.5875840189983137
torch.logcumsumexp(a) a.numel() == 100 for 20000 times torch.half
1.681279803000507
torch.logcumsumexp(a) a.numel() == 100 for 20000 times torch.float
1.1775737119969563
torch.logcumsumexp(a) a.numel() == 100 for 20000 times torch.double
6.289237987002707
```

## CPU Timings

```python
import timeit

for n, t in [(50, 20000),
             (100, 20000)]:
    for dtype in ('torch.float', 'torch.double'):
        print(f'torch.logcumsumexp(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.logcumsumexp(a, 0);torch.logcumsumexp(a, 1);',
                              setup=f'import torch; a=torch.randn(({n}, {n}), dtype={dtype}, device="cpu")',
                              number=t))
```


```
torch.logcumsumexp(a) a.numel() == 50 for 20000 times torch.float
6.126735579004162
torch.logcumsumexp(a) a.numel() == 50 for 20000 times torch.double
6.413470525003504
torch.logcumsumexp(a) a.numel() == 100 for 20000 times torch.float
24.092899453004065
torch.logcumsumexp(a) a.numel() == 100 for 20000 times torch.double
24.83073066299403
```

File: aten/src/ATen/native/cuda/ScanKernels.cu
Comment (albanD) [inline aten/src/ATen/native/cuda/ScanKernels.cu:232]: nit: variance?

**Review by albanD — COMMENTED** [review]
Looks ok, just a question on the numerical stability of the backward

File: aten/src/ATen/native/cuda/ScanKernels.cu
Comment (albanD) [inline aten/src/ATen/native/cuda/ScanKernels.cu:228]: Could you add a note here also that explains that this implementation is very close to the other scan above in this file and should be merged together to reduce duplicate code, but this is left as a future PR.

File: aten/src/ATen/native/ReduceOps.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/ReduceOps.cpp:211]: Since in the previous one, `result` is returned, while in this one we pass `result` by reference.

File: tools/autograd/templates/Functions.cpp
Comment (kshitij12345) [reply tools/autograd/templates/Functions.cpp:484]: The tests were passing. However I am not very knowledgeable about the same, so would be great if you or someone can let me know if there is a better way to go about it.

Comment (albanD) [reply tools/autograd/templates/Functions.cpp:484]: Usual tests we have use values within very small range, so very unlikely to stress this.

Could you add a new test that does the following next to the one your already added please:
- Run the forward with a Tensor of input values in larger range. For example: `a = torch.rand(10); a[0] = 100`. You can check that using `torch.logsumexp()` or your functions, this will give the right result. (You can use logsumexp() on all subsets to check your function). Note that if you try to use the ground truth you currently have in the test, you will get `inf`.
- Using the same input, you can use the `gradcheck` tool from the autograd package to check that the gradients are correct:
```
# Example for logsumexp. Same can be done for your function.
a = torch.rand(10, dtype=torch.double, requires_grad=True)
with torch.no_grad():
    a[0] = 100
gradcheck(lambda x: x.logsumexp(0), a)
```

Comment (ngimel) [reply tools/autograd/templates/Functions.cpp:484]: No need to try, self.exp() is going to kill it :-)
Does tensorflow has a formula for numerically stable backward? Otherwise someone needs to derive it.

Comment (kshitij12345) [reply tools/autograd/templates/Functions.cpp:484]: Verified. As mentioned, it blows up when it is `1000` and above.

Digged a bit into tensorflow and it looks to have a stable implementation.

https://github.com/tensorflow/tensorflow/blob/2a5910906a0e0f3dbc186ff9db6386d81a63448c/tensorflow/python/ops/math_grad.py#L1832-L1863

Will use that as a reference.

Thanks @ngimel @albanD

Comment (albanD) [reply tools/autograd/templates/Functions.cpp:484]: Perfect! Thanks for looking into it.
Ping me when you need a final review !

Comment (kshitij12345) [reply tools/autograd/templates/Functions.cpp:484]: I have tried the following which passes the gradcheck:
```python
import torch
from torch.autograd import grad
from torch.autograd import gradcheck

class LogCumsumExp(torch.autograd.Function):
    """
    We can implement our own custom autograd Functions by subclassing
    torch.autograd.Function and implementing the forward and backward passes
    which operate on Tensors.
    """

    @staticmethod
    def forward(ctx, input, dim):
        ctx.dim = dim
        with torch.no_grad():
            # Ignore Current Unstable Implementation.
            y = torch.logcumsumexp(input, ctx.dim)
        ctx.save_for_backward(input, y)
        return y

    @staticmethod
    def backward(ctx, grad_output):
        x, y = ctx.saved_tensors
        dim = ctx.dim

        grad_min = torch.tensor(torch.finfo(grad_output.dtype).min, dtype=grad_output.dtype)
        log_grad_positive = torch.where(
            grad_output > 0,
            grad_output.log(),
            grad_min)

        log_grad_negative = torch.where(
            grad_output < 0,
            (-grad_output).log(),
            grad_min)

        def reverse_logcumsum(x):
            inter = torch.logcumsumexp(torch.flip(x, (dim,)), dim)
            return torch.flip(inter, (dim,))

        output_pos = torch.exp(reverse_logcumsum(log_grad_positive - y)+ x)
        output_neg = torch.exp(reverse_logcumsum(log_grad_negative - y) + x)

        return output_pos - output_neg, None

# Input
x = torch.tensor([[10000., 0.5, 0.3], [1000., 0.5, 0.3]], dtype=torch.double)
x.requires_grad = True

logcumsumexp = LogCumsumExp()

axis = 0
gradcheck(lambda x: logcumsumexp.apply(x, axis), (x, ))

axis = 1
gradcheck(lambda x: logcumsumexp.apply(x, axis), (x, ))
```

However I have a doubt on how to go about 
```
grad_min = torch.tensor(torch.finfo(grad_output.dtype).min, dtype=grad_output.dtype)
```
in the C++ Gradient code.

Thanks.

Comment (ngimel) [reply tools/autograd/templates/Functions.cpp:484]: std::numeric_limits<scalar_t> would give you same constants that torch.finfo does.

Comment (kshitij12345) [reply tools/autograd/templates/Functions.cpp:484]: Thanks for the quick reply.
Where can I get the `scalar_t` or would we need to add a dispatch based on `scalar_t` for double and float?

File: tools/autograd/templates/Functions.cpp
Comment (kshitij12345) [inline tools/autograd/templates/Functions.cpp:508]: @ngimel @albanD 
Not sure if should use the type dispatcher this way (with `return`) as I haven't seen such use before in the codebase.
But this does work and have ran the test successfully on local.

Do let me know if there is another way to go about it.

**Comment by kshitij12345** [comment]
@ngimel @albanD Gentle ping :).

**Review by albanD — APPROVED** [review]
Looks good to me.
The perf for the backward might not be great but it is better than nothing.
Feel free to open a new issue if you think the backward should be implemented with a special kernel to make it more efficient. But I think this should be left for a future PR anyway.

**Comment by kshitij12345** [comment]
@albanD 
Would be great if this lands soon into master as it would potentially unblock #36458 .
Thank You.

**Comment by albanD** [comment]
The landing is in progress. But there are some flaky internal tests so I had to re-run these...

**Comment by agadetsky** [comment]
@kshitij12345 thank you very much, great work!

**Comment by agadetsky** [comment]
@kshitij12345 seems like documentation is a bit broken. At least computation formula does not appear at the moment at the master doc (https://pytorch.org/docs/master/generated/torch.logcumsumexp.html#torch.logcumsumexp)

**Comment by kshitij12345** [comment]
@agadetsky Thanks for notifying. Will try to get it fixed.

**Pull request closed without merging**
